engine/consensus/dkg/reactor_engine.go

```diff
@@ -88,49 +122,85 @@ func (e *ReactorEngine) Done() <-chan struct{} {
 	return e.unit.Done()
 }
 
-// EpochSetupPhaseStarted handles the EpochSetupPhaseStared protocol event. It
-// starts a new controller for the epoch and registers the triggers to regularly
-// query the DKG smart-contract and transition between phases at the specified
-// views.
+// EpochSetupPhaseStarted handles the EpochSetupPhaseStarted protocol event by
+// starting the DKG process.
 func (e *ReactorEngine) EpochSetupPhaseStarted(currentEpochCounter uint64, first *flow.Header) {
-	firstID := first.ID()
+	e.startDKGForEpoch(currentEpochCounter, first)
+}
 
-	lg := e.log.With().
+// EpochCommittedPhaseStarted handles the EpochCommittedPhaseStarted protocol
+// event by checking the consistency of our locally computed key share.
+func (e *ReactorEngine) EpochCommittedPhaseStarted(currentEpochCounter uint64, _ *flow.Header) {
+	e.handleEpochCommittedPhaseStarted(currentEpochCounter)
+}
+
+// startDKGForEpoch starts the DKG instance for the given epoch, only if we have
+// never started the DKG for the given epoch. This allows consensus nodes which
+// boot from a state snapshot within the EpochSetup phase to run the DKG.
+//
+// It starts a new controller for the epoch and registers the triggers to regularly
+// query the DKG smart-contract and transition between phases at the specified views.
+//
+func (e *ReactorEngine) startDKGForEpoch(currentEpochCounter uint64, first *flow.Header) {
+
+	firstID := first.ID()
+	nextEpochCounter := currentEpochCounter + 1
+	log := e.log.With().
 		Uint64("current_epoch", currentEpochCounter).
 		Uint64("view", first.View).
 		Hex("block", firstID[:]).
 		Logger()
 
+	// if we have started the dkg for this epoch already, exit
+	started, err := e.dkgState.GetDKGStarted(nextEpochCounter)
+	if err != nil {
+		// unexpected storage-level error
+		log.Fatal().Err(err).Msg("could not check whether DKG is started")
+	}
+	if started {
+		log.Warn().Msg("DKG already started, skipping starting the DKG for this epoch")
```

Comment:
This is the case where DKG was started in previous startup and didn't get completed before the node stopped, right? 

In that case, I changed the message a bit, because the `already` sounds like it was started by other process in the current startup.
```suggestion
		log.Warn().Msg("DKG started before, skipping starting the DKG for this epoch")
```

Reply:
Another comment is that, should we check whether we have the private key for this epoch, and verify it if we have, and insert a `nil` if we don't have? Otherwise, we will always hit a not found when retrieving the private key, which would be an exception when this node is to sign a block. 

With consensus V2, we will allow the node to sign block with staking key if DKG key is not generated, but only if a `nil` record is saved in the database. [If the key is not found, it is treated as exception](https://github.com/onflow/flow-go/pull/1194/files#diff-4ba3e824602cc8ebd30e84272cef36403ca1806f7286f34a88dc2045ada665d8R91-R92). 

So when DKG was interrupted after start, the key store gets "dirty" as no record is saved to reflect the failure, and we need to figure out when to clean it up. IMO, we should clean it here. 

Thoughts?

Reply:
> This is the case where DKG was started in previous startup and didn't get completed before the node stopped, right?

Yep. Thank you, this message is clearer

> So when DKG was interrupted after start, the key store gets "dirty" as no record is saved to reflect the failure, and we need to figure out when to clean it up

Good point, we could do it either here or when we check our result at the end of the `EpochCommitted` phase. I agree here (earlier) is a better place to do it, at least for the time being.

Reply:
We should do both I think. 

If we crash in setup phase, and start in committed phase, then we should also clean up.

But yeah, for now we can just add it here.

Reply:
cc @tarakby